### PR TITLE
lib: os: printk hook cleanups

### PR DESCRIPTION
**Pull request by fabiobaltieri**
Make `_char_out` static, cleanup the console driver code.

Will follow up on `__stdout_hook_install` separately (https://github.com/zephyrproject-rtos/zephyr/pull/76234).

### CODE REVIEW

**Review by andyross — DISMISSED** [review]
Seems clean, and a reasonable step in the direction of squaring our weird knot of kernel output primitives.  We have printk, console and logging.  And most of them can be directed at others via asymmetric means, but not in a simple way.  In the longer term it might be nice if you could have a generic "handle character output" function and use it as any/all of:

* printk hook
* Console stdout hook
* Log backend

Likewise, the decision of "where to point logging" and "where to point printk" should use the same terminology and device/API naming.  Right now you can have CONFIG_LOG_PRINTK (means "implement printk in terms of logging") which makes the printk hook a noop.  Also CONFIG_LOG_MINIMAL (means "implement logging in terms of printk"), which then relies on someone else setting up printk.

And similarly, the printf() call[1] in hello_world often refuses to work on some platforms /configs, because the console isn't set up, as everyone using those platforms doesn't care as they all do logging or printk.

None of which has anything to do with this PR which seems clean.  But something to think about if you're in the space.

[1] It used to be a printk().  Not sure when it changed but I consistently and reliably trip over the distinction.

**Comment by fabiobaltieri** [comment]
Thanks for the summary Andy - yes I bumped into this because I was looking into  figuring out the various interaction between logging with backends, logging minimal, logging with the shell backend etc... there's few combinations, not all of which necessarily lead to useful results. :-) And for the console it's odd that you can select multiple backends and they race for which one is the last setting the hook, was considering to convert the hooks into an iterable section, but then bumped into the early printk stuff as well, figured I'd just clean the headers for now.

This whole thing is a bit overwhelming for what it does.

File: drivers/console/efi_console.c [hunk 1]
```diff
@@ -16,7 +16,7 @@
 #include <zephyr/device.h>
 #include <zephyr/init.h>
 #include <zephyr/sys/printk.h>
-
+#include <zephyr/printk-hooks.h>
```
[the remark this answers is not in the document]
Comment (fabiobaltieri) [reply drivers/console/efi_console.c:19]: I really just copied this from `zephyr/sys/libc-hooks.h`. Sholud I move it under `sys`? I'll move it under `sys/`.

File: include/zephyr/printk-hooks.h [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+#define ZEPHYR_INCLUDE_PRINTK_HOOKS_H_
+
+/**
+ * @brief Install the character output routine for printk
+ *
+ * To be called by the platform's console driver at init time. Installs a
+ * routine that outputs one ASCII character at a time.
+ * @param fn putc routine to install
+ */
+void __printk_hook_install(int (*fn)(int c));
```
Comment (andyross) [inline include/zephyr/printk-hooks.h:17]: One nitpick:  if you're promoting this to an API a `typedef int (*printk_hook_fn_t)(int c);` might look cleaner than a bare function pointer.

**Comment by fabiobaltieri** [comment]
moved the header under `sys/`

File: include/zephyr/printk-hooks.h (hunk 2 above)
Comment (fabiobaltieri) [reply include/zephyr/printk-hooks.h]: Not a bad idea, added a patch for that.

**Comment by fabiobaltieri** [comment]
- added a typedef for the hook argument

**Review by andyross — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by keith-zephyr — APPROVED** [review]
Typo in the last commit message `insteadl`.

**Comment by fabiobaltieri** [comment]
> Typo in the last commit message `insteadl`.

Thanks, fixed.

**Review by aescolar — APPROVED** [review]

**Pull request merged**